gateway/src/lib.rs

```diff
@@ -151,6 +154,12 @@ impl Server {
         let mut http_servers = HashMap::with_capacity(args.addresses.len());
         let all_servers_shutdown = FuturesUnordered::new();
 
+        let metrics =
+            metrics::Metrics::new(&log, &args, config.metrics, apictx.clone())
+                .map_err(|err| {
+                    format!("failed to initialize metrics subsystem: {err}")
+                })?;
```

Comment:
Yeah, I'm honestly not sure what we ought to do here. Most of the errors
returned by `Metrics::new` are e.g. errors with the config file, which will
persist if the process is restarted. So, we probably shouldn't let MGS get into
a crash-loop if someone provides a bad config file...

Reply:
Yeah, that sounds right. Maintenance mode for MGS is probably going to make the rest of the control plane very sad! I would not block startup on this. First, as you said, it's unlikely to happen in a production environment. But when it does, we want the system to come up _somewhat_, and sacrifice the ability to produce metrics. You could also potentially keep trying to do this in a retry loop, but in a way that doesn't prevent the main server from starting?

Reply:
I also wondered if it was worth having some kind of retry loop here. However, the only errors that can cause this to fail are:
 - we didn't like the values in the config file, or
 - the call to `ProducerRegistry::register_producer` failed (which, AFAICT, only happens if we try to register the same UUID twice?)

Neither of those will go away on a retry, unless we re-load the config file from the filesystem every time. Presently, the metrics config lives in the same config file as everything else, which only gets loaded once, so it seems a *bit* weird to reload the whole file, but only respond to changes in the metrics-specific config. 

We could, alternatively, have a totally separate config file for the metrics stuff. Which, although  I don't really want to do it, *could* actually be a  good idea. Currently, since we try to parse the whole MGS config from one file, a typo or otherwise unparseable metrics config will prevent the whole gateway service from starting. Perhaps the metrics  stuff _should_ be configured  by a different file, so that screwing it up doesn't stop MGS from starting...

Or... we could just go back to hard-coding all the configuration values for metrics, so that there isn't an opportunity to make typos that can stop the management network from coming up at all.

What do you think? I'd love to get @jgallagher's opinion here as well.

Reply:
It also occurred to me that maybe we should eliminate all potential panics from the metrics code, even the ones that "should never happen" --- they could be turned into debug only panics, and the metrics subsystem could just stop running in the event of those errors in release mode. Again, with the intent to ensure that bugs in the metrics code (which is not in the critical path for the rack being "basically usable") can't take down the whole management network.

Reply:
I know we all love `panic="abort"`  around here, but...maybe this is a good argument for not aborting on panics? We could, potentially, catch all unwinds at the boundary of the metrics subsystem, and have the rest of MGS just handle them by going "welp, the metrics module has crashed, i'll log an error but choose not to die as a result of it."

Reply:
It might be worth distinguishing these two kinds of failures:

- we can't start the metrics system, for the kinds of reasons you enumerate
- some operation inside the metrics system panics

It's true that the first situation is unlikely. That means we've done something like shipped a bad configuration file, or happen to have used a duplicate UUID. Both seem unlikely, but the first doesn't sound impossible to me. It seems at least plausible that we accidentally use an invalid config file in an update -- and since we need MGS to update our way out of that problem, we shouldn't fail here!

However, the second case seems different. If we're really using panics judiciously to mean, "a condition from which I am unable or unwilling to recover has occurred," panicking seems like the right call. For one thing, these kinds of errors seem unlikely to land us in maintenance mode, because they're more likely to be transient. I could be wrong here, and a very conservative approach might be warranted given MGS's centrality to fixing many problems!

Reply:
Yup, I went ahead and ripped out the config file in 7f552778a1af02577dbe719a63cbe601205e5e9e, and I tried to clean up panics a little bit in 64ae3a0baba38ee04e9aa40457bffb6d6b2b0bc0. There's a part of me that still wants to be catching *any* unwind (even, e.g., a panic in the standard library) at the boundary of the metrics subsystem, but I don't know if that's worth it.

Reply:
FWIW, I wouldn't lose any sleep over this bit:

> * we didn't like the values in the config file

for a couple reasons specific to how this gets deployed:

* The config file is baked into the switch zone, and the switch zone is baked into the OS image. There is no mechanism (currently) where the config file contents can change without an OS update.
* If the polling intervals in the config file are bad, that means they're bad for every MGS using that Omicron build, which would be caught more or less immediately in any testing.

That said, since in practice changing the values (currently) requires an OS update, baking them into the binary is actually not any less flexible in production. If it's more annoying for dev / testing I'd be fine with going back to a config file, but probably we don't need to tweak the values for dev / testing much?

Reply:
> I tried to clean up panics a little bit in [64ae3a0](https://github.com/oxidecomputer/omicron/commit/64ae3a0baba38ee04e9aa40457bffb6d6b2b0bc0). There's a part of me that still wants to be catching _any_ unwind (even, e.g., a panic in the standard library) at the boundary of the metrics subsystem, but I don't know if that's worth it.

I've hemmed and hawed over this kind of thing myself. We've seen MGS panic unexpectedly due to a full disk, which it definitely should not (both panics were actually in dependencies; we [fixed the one in dropshot](https://github.com/oxidecomputer/dropshot/issues/811) but AFAIK the [one in slog is still an issue](https://github.com/oxidecomputer/omicron/issues/4352)). I think when push comes to shove I remain in the "panics are for broken invariants etc where a failure means we're in an unexpected state and cannot proceed", and any panics we see in practice that _aren't_ that (e.g., the panics on a full disk as above) should be changed to not panic. 

The case you're in seems particularly tricky; _even if_ there are broken invariants in the metrics system, "don't bring down the rest of MGS" seems like a reasonable take since those invariants are extremely unlikely to affect anything else, so I'm on board with the changes in https://github.com/oxidecomputer/omicron/commit/64ae3a0baba38ee04e9aa40457bffb6d6b2b0bc0 for practical reasons (even though the purist in me would vote to change those back to panics).